### ISSUE TITLE
Function bodies are not lexically scoped

### ISSUE THREAD

**Issue by vthriller** [PRQL/prql#3175]
### What's up?

```
let f = x -> (x+1)
# or: let f = x -> s"f({x})"
from foo
select (f x)
```

```
Error: 
   ╭─[:1:15]
   │
 1 │ let f = x -> (x+1)
   │               ┬  
   │               ╰── Ambiguous name
   │ 
   │ Help: could be any of: _param.x, this.foo.x
───╯
```

Gladly there's this "ambiguous name" error, but that's not typo-proof. Misspell an argument and, voilà, you're calculating over a table field you didn't want to touch. It would be more difficult should PRQL become schema-aware; asking users to prefix everything with `_param.` in such case would be like asking them to not overrun buffers, but with less drastic consequences: forget one instance, and users of your module are complaining that function doesn't work with their derived field.

I'm pretty sure most would prefer and/or expect lexical scoping for functions. The question is whether people would also want proper macros, and how useful such macros would actually be.

**Comment by max-sixty** [comment]
At the moment this is a deliberate choice (though not sure it's necessarily the correct one)

With lexical scoping, it's not possible to easily "templatize" — for example, from https://github.com/PRQL/dbt-prql, with old syntax:

```
func filter_amount method -> amount | filter payment_method == method | sum

from {{ ref('raw_payments') }}
group order_id (
  aggregate [
    bank_transfer_amount = filter_amount 'bank'
    credit_card_amount   = filter_amount 'credit_card'
    gift_amount          = filter_amount 'gift_card'
  ]
)
```

Open to ideas with examples for the tradeoff...

**Comment by vthriller** [comment]
I see. I wasn't sure whether use cases for proper macros that came to mind were niche enough to be discarded outright.

> Note that when https://github.com/prql/prql/issues/82 is implemented

Speaking of unimplemented macros, here's that hypothetical example I was thinking about:

```
let join_user = field name t -> (t | join $name = users ($name.id==$field))

from i = issues
join_user i.assignee a
join_user i.reporter r
filter (! a.enabled) || (! r.enabled)
```

Although I'm pretty sure this particular use case would rather benefit from first-class support of some sort.

**Comment by max-sixty** [comment]
Yeah — IIUC — macros which enable prql variable names to be templated is probably out of scope in the medium term, in order to keep the language simple...

Let's leave this issue open re lexical scoping though — that does seem like a useful area of discussion
